### PR TITLE
CONTRIBUTING: Use PEP 604 syntax

### PR DESCRIPTION
**Pull request by srittau**
Built-in generics not possible in type aliases

### CODE REVIEW

**Comment by hoefling** [comment]
@srittau it's also not possible to use PEP 604 syntax in class bases, e.g.
```py
class c_char_p(_PointerLike, _SimpleCData[bytes | None]): ...
```
instead of
https://github.com/python/typeshed/blob/a6ce6fdce600cfd2faf3910d3a7331411e378279/stdlib/ctypes/__init__.pyi#L218

**Comment by srittau** [comment]
> @srittau it's also not possible to use PEP 604 syntax in class bases, e.g.

Good catch, updated.

**Comment by Akuli** [comment]
I didn't do this in #5879, because I thought it would be easier to say "Use the new syntax", and when needed, say "mypy is still buggy, use the old syntax here" when reviewing.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -327,9 +327,10 @@ checker, and leave out unnecessary detail:
 
 Some further tips for good type hints:
 * use built-in generics (`list`, `dict`, `tuple`, `set`), instead
-  of importing them from `typing`, **except** for arbitrary length tuples
-  (`Tuple[int, ...]`) (see
-  [python/mypy#9980](https://github.com/python/mypy/issues/9980));
+  of importing them from `typing`, **except** in type aliases and for
+  arbitrary length tuples (`Tuple[int, ...]`);
+* use `X | Y` instead of `Union[X, Y]` and `X | None`, instead of
+  `Optional[X]`, **except** in type aliases and class bases;
```
Comment (Akuli) [inline CONTRIBUTING.md:333]: Would be good to mention why this rule exists:

```suggestion
  `Optional[X]`, **except** when it is not possible due to mypy bugs (type aliases and base classes);
```

**Comment by hoefling** [comment]
FWIW, both issues with PEP 604 syntax are resolved on `mypy`'s master, so the 0.920 / 0.1000 /whatever should obsolete the warnings.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -327,9 +327,10 @@ checker, and leave out unnecessary detail:
 
 Some further tips for good type hints:
 * use built-in generics (`list`, `dict`, `tuple`, `set`), instead
-  of importing them from `typing`, **except** for arbitrary length tuples
-  (`Tuple[int, ...]`) (see
-  [python/mypy#9980](https://github.com/python/mypy/issues/9980));
+  of importing them from `typing`, **except** in type aliases and for
+  arbitrary length tuples (`Tuple[int, ...]`);
```
Comment (Akuli) [inline CONTRIBUTING.md:331]: Apparently base classes didn't make it to the previous line. I don't think inheriting from `tuple[int, ...]` works.

```suggestion
  of importing them from `typing`, **except** in type aliases, in base classes, and for
  arbitrary length tuples (`Tuple[int, ...]`);
```

**Pull request merged**